### DISCUSSION TITLE
Use UploadFile in Pydantic model

### DISCUSSION THREAD

**Discussion by laeubli** [fastapi/fastapi#7964] [Questions] [answered]
**Description**

Is it possible to use `UploadFile` in a Pydantic model? The FastAPI docs [say](https://fastapi.tiangolo.com/tutorial/request-files/) “FastAPI's UploadFile inherits directly from Starlette's UploadFile, but adds some necessary parts to make it compatible with Pydantic and the other parts of FastAPI.” 

```python3
from fastapi import FastAPI, File, UploadFile
from pydantic import BaseModel

class PydanticFile(BaseModel):
    file: UploadFile = File(...)

app = FastAPI()

@app.post("/uploadfile/")
async def create_upload_file(file: PydanticFile):
    return {"filename": file.filename}
```

However, `UploadFile` does not seem to be compatible with Pydantic. The code above – derived from the example in FastAPI's [Request Files tutorial](https://fastapi.tiangolo.com/tutorial/request-files/) – raises a value error:

```
ValueError: Value not declarable with JSON Schema, field: file type=UploadFile required
```

**Comment by michaelschmit** [comment]
I'm a FastAPI/Starlette noob so take what I'm saying with a grain of salt, but I don't think what you are trying to do really makes sense.

When you are specifying a UploadFile I believe you are telling FastAPI/Starlette that you will be providing data in a multipart form body.  Pydantic is for describing the expected JSON format body.  So trying to combine the two doesn't really make sense because you need to do one or the other.

I could be off base though ... you can wait for someone else to chime in.

**Comment by laeubli** [comment]
Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:

```python3
class Properties(BaseModel):
    language: str = None
    author: str = None

@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile = File(...),
                             properties: Properties):
    return {"filename": file.filename}
```

This seems to be working, and maybe query parameters would ultimately make more sense here.

**Comment by dmontagu** [comment]
When it says "adds necessary parts to make it compatible with Pydantic", it doesn't mean with pydantic BaseModel instances. FastAPI makes deep use of a lot of pydantic internals during the request-handling process, and I believe the modifications to the type are to make it compatible with that. In particular, it is used in a way that doesn't require it to be able to live as a field on a `BaseModel`.

Generally, you should only use `BaseModel` instances in FastAPI when you know you want to parse the model contents from the json body of the request.

If you want to make use of `UploadFile` as an attribute of a dependency class you can, it just can't be a pydantic model. (You can look at how the auth dependencies are implemented for an example along those lines.) A plain-old 3.7 dataclass would probably work for this.

**Comment by michaelschmit** [comment]
> This seems to be working, and maybe query parameters would ultimately make more sense here.

I would go the more RESTful approach and use /author/{author name or id}/language/{language name or id}/file.  Upload is kind of implied with the POST request.  With that said, this discussion is no long Fast API based so we should probably discontinue that discussion here.

Good day and good luck ...

**Comment by laeubli** [comment]
Thanks a lot @michaelschmit and @dmontagu.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue @laeubli  :+1:

**Comment by perezzini** [comment]
Hello,

So, what is the best approach when working with request bodies (represented as Pydantic models) and `UploadFiles` in FastAPI? How could I send the request body through `multipart/form-data`?

I'm pretty sure I am missing concepts about general HTTP requests, but if someone has a quick answer It'd be awesome.

Thanks in advance.

**Comment by divyeshwendorse** [comment]
> Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:
> 
> ```python
> class Properties(BaseModel):
>     language: str = None
>     author: str = None
> 
> @app.post("/uploadfile/")
> async def create_upload_file(file: UploadFile = File(...),
>                              properties: Properties):
>     return {"filename": file.filename}
> ```
> 
> This seems to be working, and maybe query parameters would ultimately make more sense here.

Hello,
This is giving an validation Error (422)

**Comment by perezzini** [comment]
> > Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:
> > ```python
> > class Properties(BaseModel):
> >     language: str = None
> >     author: str = None
> > 
> > @app.post("/uploadfile/")
> > async def create_upload_file(file: UploadFile = File(...),
> >                              properties: Properties):
> >     return {"filename": file.filename}
> > ```
> > 
> > 
> > This seems to be working, and maybe query parameters would ultimately make more sense here.
> 
> Hello,
> This is giving an validation Error (422)

That's right. That example is not valid.

**Comment by tiangolo** [comment]
@perezzini if you are receiving JSON data, with `application/json`, use normal Pydantic models.

This would be the most common way to communicate with an API.

If you are receiving a raw file, e.g. a picture or PDF file to store it in the server, then use `UploadFile`, it will be sent as form data (`multipart/form-data`).

If you need to receive some type of structured content that is not JSON but you want to validate in some way, for example, an Excel file, you would still have to upload it using `UploadFile` and do all the necessary validations in your code. You could use Pydantic in your own code for your validations, but there's no way for FastAPI to do it for you in that case.

**Comment by perezzini** [comment]
Thanks for replying back @tiangolo 

And congrats on your great job in FastAPI!

**Comment by rohansaw** [comment]
> > > Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:
> > > ```python
> > > class Properties(BaseModel):
> > >     language: str = None
> > >     author: str = None
> > > 
> > > @app.post("/uploadfile/")
> > > async def create_upload_file(file: UploadFile = File(...),
> > >                              properties: Properties):
> > >     return {"filename": file.filename}
> > > ```
> > > 
> > > 
> > > This seems to be working, and maybe query parameters would ultimately make more sense here.
> > 
> > 
> > Hello,
> > This is giving an validation Error (422)
> 
> That's right. That example is not valid.

Hi, Did you find a solution to pass UploadFile and a custom PydanticModel to a Request? I' currently stuck with the same validation Error (422)

**Comment by alejandrojcastaneira** [comment]
Hello everyone, 

I would also like to find a solution for this, my use case will be to send documents to a ML model in production and make predictions over those. But the input could vary between document files such as .pdf, .word, etc. and also there is a possibility to have plane text within a json. 

Ideally, I would love a query parameter named `input` to switch between these types, so I wouldn't have to declare different endpoints for json or documents. But at the moment I can't send both a UploadFile and a custom PydanticModel inside a single request.

Any help would be greatly! appreciated.

Best regards

**Comment by ar4hc** [comment]
Came here from Goggle, and i found this on [SO](https://stackoverflow.com/questions/60127234/fastapi-form-data-with-pydantic-model) which could help solve this.

**Comment by amjltc295** [comment]
> > > > Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:
> > > > ```python
> > > > class Properties(BaseModel):
> > > >     language: str = None
> > > >     author: str = None
> > > > 
> > > > @app.post("/uploadfile/")
> > > > async def create_upload_file(file: UploadFile = File(...),
> > > >                              properties: Properties):
> > > >     return {"filename": file.filename}
> > > > ```
> > > > 
> > > > 
> > > > This seems to be working, and maybe query parameters would ultimately make more sense here.
> > > 
> > > 
> > > Hello,
> > > This is giving an validation Error (422)
> > 
> > 
> > That's right. That example is not valid.
> 
> Hi, Did you find a solution to pass UploadFile and a custom PydanticModel to a Request? I' currently stuck with the same validation Error (422)

I had the same problem. Unfortunately, you cannot do that. It's a HTTP limitation. See https://github.com/tiangolo/fastapi/issues/285#issuecomment-498368734

**Comment by perezzini** [comment]
> > > > > Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:
> > > > > ```python
> > > > > class Properties(BaseModel):
> > > > >     language: str = None
> > > > >     author: str = None
> > > > > 
> > > > > @app.post("/uploadfile/")
> > > > > async def create_upload_file(file: UploadFile = File(...),
> > > > >                              properties: Properties):
> > > > >     return {"filename": file.filename}
> > > > > ```
> > > > > 
> > > > > 
> > > > > This seems to be working, and maybe query parameters would ultimately make more sense here.
> > > > 
> > > > 
> > > > Hello,
> > > > This is giving an validation Error (422)
> > > 
> > > 
> > > That's right. That example is not valid.
> > 
> > 
> > Hi, Did you find a solution to pass UploadFile and a custom PydanticModel to a Request? I' currently stuck with the same validation Error (422)
> 
> I had the same problem. Unfortunately, you cannot do that. It's a HTTP limitation. See [#285 (comment)](https://github.com/tiangolo/fastapi/issues/285#issuecomment-498368734)

Yes, it's a HTTP limitation; not an issue regarding FastAPI.

**Comment by synchronizing** [comment]
For anyone else wondering, a solution was posted on #2257.

**Comment by MuhammadAbdullah54321** [comment]
> @perezzini if you are receiving JSON data, with `application/json`, use normal Pydantic models.
> 
> This would be the most common way to communicate with an API.
> 
> If you are receiving a raw file, e.g. a picture or PDF file to store it in the server, then use `UploadFile`, it will be sent as form data (`multipart/form-data`).
> 
> If you need to receive some type of structured content that is not JSON but you want to validate in some way, for example, an Excel file, you would still have to upload it using `UploadFile` and do all the necessary validations in your code. You could use Pydantic in your own code for your validations, but there's no way for FastAPI to do it for you in that case.

But in case of image, how will you send image from client side if we will not use pydantic model ??

**Comment by VpkPrasanna** [comment]
Is it possible some thins like this ?
>
>```python
>class ocrDetection(BaseModel):
>    URL:Optional[str]
>    Image:Optional[UploadFile] = File(...)
```
to achieve using pydantic?

**Comment by VpkPrasanna** [comment]
> Thanks a lot for your helpful comment. So I guess I'd have to explicitly separate the file from the [JSON part of the multipart form body](url), as in:
> 
> ```python
> class Properties(BaseModel):
>     language: str = None
>     author: str = None
> 
> @app.post("/uploadfile/")
> async def create_upload_file(file: UploadFile = File(...),
>                              properties: Properties):
>     return {"filename": file.filename}
> ```
> 
> This seems to be working, and maybe query parameters would ultimately make more sense here.

If yo do so both will be required even though you have provided with optional in Properties Class it is mandatory, but i need something like both has to be optional,any help. 
TIA

**Comment by bharath-kumarn** [comment]
error:  non-default argument follows default argument


    async def create_upload_file(file: UploadFile = File(...),properties: Properties):
                                 ^
SyntaxError: non-default argument follows default argument

**Comment by richylyq** [comment]
> error: non-default argument follows default argument
> 
> ```
> async def create_upload_file(file: UploadFile = File(...),properties: Properties):
>                              ^
> ```
> 
> SyntaxError: non-default argument follows default argument

hi, solution to this issue is just something like this
`async def create_upload_file(properties: Properties, file: UploadFile = File(...)):`

**Comment by AvaiyaSoftware** [comment]
@richylyq 
it's returning 422.

`422Unprocessable Entity`


```
{
    "detail": [
        {
            "loc": [
                "body",
                "optimizationScenario"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        },
        {
            "loc": [
                "body",
                "plannedMaintenance"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        }
    ]
}
```

**Comment by AvaiyaSoftware** [comment]
Yes, it's a HTTP limitation; not an issue regarding FastAPI.

**Comment by kunansy** [comment]
FastAPI can't handle UploadFile and JSON body together, it's very inconvenient.
E.g. to get data from HTML form and send email with attached files.

Any solutions? I have to get them, but can't 😞

**Comment by AvaiyaSoftware** [comment]
Formdata

**Comment by rudy-haryanto** [comment]
> Formdata
> […](#)
> On Thu, 25 Nov, 2021, 3:34 PM Kirill, ***@***.***> wrote: FastAPI can't handle UploadFile and JSON body together, it's very inconvenient. E.g. to get data from HTML form and send email with attached files. Any solutions? I have to get them, but can't 😞 — You are receiving this because you commented. Reply to this email directly, view it on GitHub <[#657 (comment)](https://github.com/tiangolo/fastapi/issues/657#issuecomment-979052049)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/APWBSLXOO53LVQVMFKKZZKTUNYCY7ANCNFSM4JFI7ZBA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

Can you please elaborate more? I had this same issue and still getting the 422 with above code.

**Comment by laleet-avaiya** [comment]
@rudy-haryanto 

https://www.freecodecamp.org/news/formdata-explained/

And same accept it on FastApI side.

**Comment by spaceofmiah** [comment]
@laeubli I wrote an article to address this issue and also implemented a solution that works with fastapi

https://dev.to/spaceofmiah/api-file-upload-done-right-fastapi-1kd1#how-to-handle-file-upload-with-pydantic-model

**Reply by louis-she** [reply]
thanks for the tip, but it only works for small files, in the case of large file the memory will be eaten up with base64, we do need a solution work with `UploadFile`.

**Comment by louis-she** [comment]
i really do not want 2 same validation code, my work around is

0. user click submit
1. post meta data to create the record in db.
2. upload file and assiciate the file to that record.
3. alert user the submition successed.

hope that helps.

**Comment by prannoytank** [comment, marked as answer]
```
from pydantic import BaseModel, StringConstraints, EmailStr,model_validator
from fastapi import UploadFile, File, Form
from fastapi.responses import JSONResponse

class RequestData(BaseModel):
    email: Annotated[EmailStr, StringConstraints(max_length=255)]
    password: Annotated[str, StringConstraints(max_length=60)]
    sample_file: Annotated[UploadFile, File()]
    model_config = {"extra": "forbid"}

    @model_validator(mode='after')
    def check_file_type_validation(self) -> Self:
        if self.sample_file.content_type != 'application/pdf':
            raise ValueError('File Type not supported')

        if self.sample_file.size > 5000:
            raise ValueError(f"Max file size allowed : 5000 bytes")

        return self

@router.post("/multi_part_form_data")
def multi_part_form_data(
        request_data: RequestData = Form(..., media_type="multipart/form-data")):  # Pydantic request body via Form

    return {
        "file_name": request_data.sample_file.filename,
        "email": request_data.email,
        "password": request_data.password
    }
	

Sample Curl:
   
curl -X 'POST' \
  'http://localhost:8000/test/multi_part_form_data' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'email=user@example.com' \
  -F 'password=sample_password' \
  -F 'sample_file=@sample_file.txt;type=text/plain'

```
